### PR TITLE
gpu: sycl: add reorder primitive

### PR DESCRIPTION
**Pull request by t4c1**
Adds SYCL implementation of reorder primitive.

### CODE REVIEW

File: src/gpu/gpu_reorder_list.cpp [hunk 1]
```diff
@@ -53,6 +54,7 @@ constexpr impl_list_item_t impl_list[] = REG_REORDER_P({
         GPU_REORDER_INSTANCE_INTEL(intel::ocl::ref_reorder_t::pd_t)    // slow but fits every use case
         GPU_REORDER_INSTANCE_NVIDIA(intel::ocl::cross_engine_reorder_t::pd_t)
         GPU_REORDER_INSTANCE_NVIDIA(nvidia::cudnn_reorder_t::pd_t)
+        GPU_REORDER_INSTANCE_NVIDIA(sycl::ref_reorder_t::pd_t)
```
Comment (dzarukin) [inline src/gpu/gpu_reorder_list.cpp:57]: I bet this should be GENERIC version instead.

File: src/gpu/sycl/ref_reorder.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+* Copyright 2022-2024 Intel Corporation
```
Comment (dzarukin) [inline src/gpu/sycl/ref_reorder.hpp:2]:
```suggestion
* Copyright 2024 Intel Corporation
```

File: src/gpu/sycl/ref_reorder.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/*******************************************************************************
+* Copyright 2022-2023 Intel Corporation
```
Comment (dzarukin) [inline src/gpu/sycl/ref_reorder.cpp:2]:
```suggestion
* Copyright 2024 Intel Corporation
```

File: src/gpu/sycl/ref_reorder.hpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        using gpu_reorder_pd_t::gpu_reorder_pd_t;
+
+        DECLARE_COMMON_PD_T("dpcpp:ref:any", ref_reorder_t);
+
+        status_t init(
+                engine_t *engine, engine_t *src_engine, engine_t *dst_engine) {
+            using namespace data_type;
+            using sm = primitive_attr_t::skip_mask_t;
+
+            const memory_desc_wrapper src_d(src_md(0));
```
Comment (dzarukin) [inline src/gpu/sycl/ref_reorder.hpp:46]: What's the purpose of explicit `0`?

File: src/gpu/sycl/reorder_kernels.hpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
+/*******************************************************************************
+* Copyright 2022-2023 Intel Corporation
```
Comment (dzarukin) [inline src/gpu/sycl/reorder_kernels.hpp:2]:
```suggestion
* Copyright 2024 Intel Corporation
```

File: tests/benchdnn/reorder/reorder.cpp [hunk 6]
```diff
@@ -146,6 +146,18 @@ dnnl_status_t init_pd(init_pd_args_t<prb_t> &init_pd_args) {
...
+        for (int i = 0; i < prb->ndims; i++) {
+            if (prb->dims[i] > INT_MAX) {
+                BENCHDNN_PRINT(2,
+                        "[CHECK_MEM]: Dimension of size %ld "
+                        "is to big to fit in an int (max %d).\n",
+                        prb->dims[i], INT_MAX);
+                return dnnl_unimplemented;
+            }
+        }
+    }
```
Comment (dzarukin) [inline tests/benchdnn/reorder/reorder.cpp:159]: This is the library limitation, not benchdnn's. Should be moved from here.

File: src/gpu/sycl/sycl_primitive_conf.hpp [hunk 7]
```diff
@@ -127,6 +127,25 @@ struct sycl_shuffle_conf_t {
...
+    bool do_scale_src;
+    int scale_src_mask;
+    bool do_scale_dst;
+    int scale_dst_mask;
+
+    int ndims;
+
+    int block_size;
+    int wg_size;
+    int wk_size;
```
Comment (dzarukin) [inline src/gpu/sycl/sycl_primitive_conf.hpp:144]: Mind to comment the usage of these members?

File: src/gpu/sycl/ref_reorder.hpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            const memory_desc_wrapper dst_d(dst_md());
+
+            const bool ok = check_data_types(src_d, dst_d)
+                    && check_formats(src_d, dst_d)
+                    && attr()->has_default_values(
+                            sm::scales_runtime | sm::post_ops)
+                    && (src_md(0)->format_desc.blocking.inner_nblks == 0)
+                    && (dst_md()->format_desc.blocking.inner_nblks == 0)
```
Comment (dzarukin) [inline src/gpu/sycl/ref_reorder.hpp:54]: Minor: better be a part of `check_formats`.

File: tests/benchdnn/reorder/reorder.cpp (hunk 6 above)
Comment (t4c1) [reply tests/benchdnn/reorder/reorder.cpp:159]: Where should it be moved to?

File: src/gpu/sycl/sycl_primitive_conf.hpp (hunk 7 above)
Comment (t4c1) [reply src/gpu/sycl/sycl_primitive_conf.hpp:144]: These were copied from binary primitive. Personally I am not a fan of how this is done and do not have that great understanding of that, but I see value of keeping implementations consistent between primitives.

File: src/gpu/gpu_reorder_list.cpp (hunk 1 above)
Comment (t4c1) [reply src/gpu/gpu_reorder_list.cpp]: Actually, after looking at what other options exist, I believe it should be GENERIC_SYCL.

File: src/gpu/sycl/ref_reorder.hpp (hunk 2 above)
Comment (t4c1) [reply src/gpu/sycl/ref_reorder.hpp]: Fixed

File: src/gpu/sycl/ref_reorder.cpp (hunk 3 above)
Comment (t4c1) [reply src/gpu/sycl/ref_reorder.cpp]: Fixed

File: src/gpu/sycl/ref_reorder.hpp (hunk 4 above)
Comment (t4c1) [reply src/gpu/sycl/ref_reorder.hpp]: None, I can remove it.

File: src/gpu/sycl/reorder_kernels.hpp (hunk 5 above)
Comment (t4c1) [reply src/gpu/sycl/reorder_kernels.hpp]: Fixed

File: src/gpu/sycl/ref_reorder.hpp (hunk 8 above)
Comment (t4c1) [reply src/gpu/sycl/ref_reorder.hpp]: This was actually a duplicate of what is checked in `check_formats`. Removed.

File: tests/benchdnn/reorder/reorder.cpp (hunk 6 above)
Comment (t4c1) [reply tests/benchdnn/reorder/reorder.cpp:159]: Do you mean `init_pd()` in the `reorder.cpp` in benchdnn?

File: src/gpu/sycl/sycl_primitive_conf.hpp (hunk 7 above)
Comment (t4c1) [reply src/gpu/sycl/sycl_primitive_conf.hpp:144]: We probably will at one point, but currently that is not high priority.

File: tests/benchdnn/reorder/reorder.cpp (hunk 6 above)
Comment (dzarukin) [reply tests/benchdnn/reorder/reorder.cpp:159]: I mean `init()` function inside the reorder implementation inside the library. It's a user-visible limitation and the implementation should dispatch into if such dimensions encountered.

Comment (t4c1) [reply tests/benchdnn/reorder/reorder.cpp]: Done

**Pull request merged**